Adding Name Encoding Improvement without MbString Extension

SirLouen opened the pull request:
More information about this here:
https://github.com/PHPMailer/PHPMailer/issues/3201

The problem is that the section of the code that is decoding those values is not tackling well charsets, specially UTF-8. There might be more scenarios untackled, as handling this without `mbstring` feels a bit arbitrary. Feel free to add more cases in a review.

SirLouen commented on src/PHPMailer.php:
```diff
@@ -1321,6 +1301,69 @@ public static function parseAddresses($addrstr, $useimap = true, $charset = self
+
+        // If no mbstring, use a manual decoder
+        $decoded = preg_replace_callback(
+            '/=\?([^?]+)\?([bqBQ])\?([^?]+)\?=/s',
```
@Synchro I'm not really sure what you mean. 
If you want to use multiple encodings at the same time, you might be using UTF-8 or even UTF-16 depending on which characters you want to use combined.

SirLouen commented:
@Synchro I'm thinking now that the selected name for the function is unfortunate (although we could consider extending it). At this moment, it's only being targeted by the `parseAddresses` function. And more specifically, the name within an address (The Name <the@address.com>) 

With Imap parsing the `$address->personal`
and without IMAP parsing: `$name`

But it could definitely be 
A) extended to cover each single Header value in the system. (I have not reviewed the other, though)
B) Maybe better changing the name to `decodeAddressNameValue`

Also I've been thinking about the tests for no-mbstring

Why the expected value is the encoded message?
```
    'name'    => '=?utf-8?B?0J3QsNC30LLQsNC90LjQtSDRgtC10YHRgtCw?=',
                            'address' => 'encoded@example.org',
```

I modified this section https://github.com/SirLouen/PHPMailer/blob/bf8e97d3c83247948ca5537585224ae8a3ab3140/test/SendTestCase.php#L68-L75
```diff
            switch ($requestKey) {
                case 'mail_to':
                    $this->propertyChanges[$phpmailerKey] = [
                        'address' => $_REQUEST[$requestKey],
-                        'name'    => 'Test User',
+                        'name'    => 'Test ⭐ User',
                    ];
                    break;
```

And this is what I'm receiving in the environment without mbstring
Test â­ User <to@example.com>

And this in the regular environment
Test ⭐ User <to@example.com>

Although my tests are passing (I even added an extra test with emojis to the `dataAddressSplitting` dataset like:

```
                        [
                            'name'    => '=?UTF-8?B?SGVsbG8g8J+MjSDkuJbnlYwgY2Fmw6k=?=',
                            'address' => 'encoded2@example.org',
                        ],
```

Conclusion: The no-mbstring tests seem to be screwed… because the code for this use-case/environment is not right

Worth the effort fixing this? As I said in 2025, I feel that `mbstring` is not an option any more, it feels very PHP 5.X.

Synchro commented:
Running without `mbstring` is not a primary aim, and support does not have to be comprehensive. Historically PHPMailer has defaulted/assumed ISO-8859-1 (as has PHP), and that's something that can be handled without mbstring, so long as you don't expect any kind of clever charset mapping. There's no point in trying to implement an encoding engine when installing mbstring solves the problem so easily. I think it's fine for it to fail in ugly ways if you provide UTF-8 content when mbstring isn't available; just *don't do that*™. While not every ISO-8859-x string is valid UTF-8, the reverse is always true; it will look bad, but that's the caller's problem.

In short, we should not try to make this decoding work without mbstring.

Synchro commented:
This is all good, and thanks for your efforts.

I didn't realise that iconv was a default extension, so it makes sense to use it if it's there, though I don't want to get into the situation of duplicating everything that mbstring does using iconv instead. We should add `ext-iconv` to the `composer.json` suggest section too. I've also seen that the imap extension is somewhat unloved, and often omitted from packaged PHP versions, hence the appearance of [imapengine](https://packagist.org/packages/directorytree/imapengine).

Regarding the fallback case, I don't think we need to care: if you want to use UTF-8, you must have mbstring. If you try to use it without, it's going to break.

On the `useimap` param, I agree; I don't quite recall why I added it at the time; the main thing was that I wanted the address parsing to be a standalone feature. If the automated choice is reliable, leave the param in place, mark it as deprecated, and ignore it internally. I deprecated `SingleTo` a while ago; that kind of thing should be done in the app layer, outside PHPMailer.

Separately, could you please sign your commits? Downstream packagers like that.

Synchro commented:
It's also failing because of [a bug in the composer-install GitHub action](https://github.com/ramsey/composer-install/issues/270) that breaks on PHP 8.5.

jrfnl commented:
> So near... Now we have [another PHP 8.5 problem](https://github.com/PHPMailer/PHPMailer/actions/runs/17121380830/job/48629457012?pr=3202) triggered by the linter.
> 
> @jrfnl Do you know how we suppress that, since it's intentional?

Yup, if it's intentional it can be ignored.

There are two/three options:
* Either ignore this file by default for the linter and have a separate lint run specifically for this file without the `--show-deprecated` flag.
* Add [a comment](https://github.com/php-parallel-lint/PHP-Parallel-Lint#excluding-files-from-a-scan-based-on-the-php-version-used) on the line containing the `<?php` tag to only lint this file on PHP < 8 .5: `<?php // lint < 8.5`.

However, as this is a security related test fixture, I would recommend splitting the test fixture file into two files - one containing everything except that one line, the other containing just that one line.

Then for the split off test fixture file with just the `ls-l` code, apply one of the above mentioned options.

As a last alternative, the backticks could be changed to `shell_exec()`.

If you let me know what you prefer, I'd be happy to send you a PR.

Synchro commented:
Yes please to the PR. I think the one-line split is the best approach, and there is already a function call test in there, so we need to retain the backticks for < 8.5.

Synchro commented on composer.json:
```diff
@@ -49,6 +49,7 @@
     },
     "suggest": {
         "decomplexity/SendOauth2": "Adapter for using XOAUTH2 authentication",
+        "ext-iconv": "To support UTF-8 if the mbstring PHP extension is not enabled",
```
This makes it sound like `iconv` is a complete replacement for `mbstring`, but I think we are only explicitly using it in `parseAddresses`, so that's misleading.

SirLouen replied:
I'm currently working on this. I will ping you out when I have this finished.

SirLouen commented:
Ok, I think I got it now

1. Added the better parsing for #3235 
2. Better ordered the tests in `ParseAddressesTest.php`
3. Removed all traces of `iconv` as commented by the WP meta team `iconv` and `mbstring` are pretty much the same % so we can get rid of so much duplicated code. 
4. Now I can take advantage of the new static translator, so I can use this for the deprecated argument
5. And now all functions have their specific test coverage.

SirLouen commented on src/PHPMailer.php:
```diff
@@ -1237,16 +1237,19 @@ protected function addAnAddress($kind, $address, $name = '')
      * @return array
      */
-    public static function parseAddresses($addrstr, $useimap = true, $charset = self::CHARSET_ISO88591)
+    public static function parseAddresses($addrstr, $deprecatedArg = null, $charset = self::CHARSET_ISO88591)
```
I think that we could perfectly roll back this little change because now we have the right notice text, that we didn't before the "static-fication" of the `setLanguage` function. When I added the first code, it felt a little confusing of why this was being deprecated without adding some change like this, but now [it's clearly advertised](https://github.com/PHPMailer/PHPMailer/blob/6ed58a9c607f1830bea0e58f88cf30f7344f8b3f/src/PHPMailer.php#L1253) if anyone tries to use this argument without using a `null`. 

What I'm not 100% clear, maybe @jrfnl knows more about this, when an expected value is no more the value in particular, in this case `bool` shall we maintain the docs typing? Docs won't provoke any errors, but they will alert the developers in their IDE (apart from the obvious runtime notice) that they should be changing their code (if they are using a bool, they will get the IDE error like: `bool` is not admitted. 

> Where does the string come from ? And shouldn't it allow for bool in case people still pass the old argument ?
Yes, the `string` is my mistake, when I was researching for a deprecation reference to apply here, I copied this from somewhere I can't remember, and I completely forgot to remove the `string` type. 

Conclusion: 
I would get back the same naming for BC, remove the `string` typing, and only leave `null` as the sole accepted value.

Makes sense @jrfnl @Synchro?
